### PR TITLE
FTE permission and camera preview

### PR DESCRIPTION
**Pull request by punamdahiya**

### CODE REVIEW

**Comment by punamdahiya** [comment]
WIP PR attempts:
- Display camera preview (Added PreviewView UIView element)
- Use TTS to announce ‘Light house app is ready’
- Handles App access to camera permission first time app is opened after install (User Story # 19)
- Entry point ‘Capture Item’ button in bottom toolbar letting app know user would like to capture new image (User Story #21)
- Current UI Accessibility and voice over compatibility

**Comment by punamdahiya** [comment]
Updated to use iOS10 AVCapturePhotoOutput

File: Lighthouse Camera/UI/Main/AlertHelper.swift [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+extension UIAlertController {
+  // Customise UIAlertController to handle permission failure alerts
+
+  enum AlertMessageID {
+    case CameraFailed, MicrophoneFailed
+  }
+
+  convenience init(messageID: AlertMessageID) {
+    switch messageID {
+    case .CameraFailed:
```
Comment (Yoric) [inline Lighthouse Camera/UI/Main/AlertHelper.swift]: I don't know of any guidelines for Swift, but I suspect that indenting the `case` would make things a bit more readable.

File: Lighthouse Camera/UI/Main/AlertHelper.swift [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+//
+//  Created by Punam Dahiya on 1/26/17.
+//  Copyright © 2017 Lighthouse. All rights reserved.
+//
+
+extension UIAlertController {
+  // Customise UIAlertController to handle permission failure alerts
+
+  enum AlertMessageID {
+    case CameraFailed, MicrophoneFailed
```
Comment (Yoric) [inline Lighthouse Camera/UI/Main/AlertHelper.swift]: Since you're asking me for a review, I'll ask you for more documentation :)
In this case, who's expected to send `CameraFailed`, `MicrophoneFailed`? And can you mention *what* failed?

File: Lighthouse Camera/UI/Main/PermissionHelper.swift [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+    case .denied, .restricted:
+      self.present(UIAlertController(messageID: .CameraFailed), animated: true, completion: nil)
+      // The user doesn't have the authority to request access e.g. parental restriction.
+      completionHandler(false)
+    }
+  }
+  // Check and grant permission for accessing the Microphone.
+
+  func checkMicrophoneAuthorization(_ completionHandler: @escaping ((_ authorized: Bool) -> Void)) {
+    AVAudioSession.sharedInstance().requestRecordPermission({ (granted: Bool) -> Void in
```
Comment (Yoric) [inline Lighthouse Camera/UI/Main/PermissionHelper.swift]: So it's entirely different results for `requestRecordPermission` and `AVCaptureDevice.authorizationStatus`? Weird API.

File: Lighthouse Camera/UI/Main/PreviewView.swift [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+
+import UIKit
+import AVFoundation
+
+protocol PreviewViewDelegate: class {
+  func displayPreviewSession()
+}
+
+
+class PreviewView: UIView {
```
Comment (Yoric) [inline Lighthouse Camera/UI/Main/PreviewView.swift]: What is it supposed to preview, exactly?

File: Lighthouse Camera/UI/Main/PreviewView.swift [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+      return NSLocalizedString("Peview Screen", comment: "")
+    }
+    set {
+      self.accessibilityLabel = NSLocalizedString("Peview Screen", comment: "")
+    }
+  }
+    
+  func startPreviewSession() {
+    // Establish Capture Session
+    let captureSession = AVCaptureSession()
```
Comment (Yoric) [inline Lighthouse Camera/UI/Main/PreviewView.swift]: So we're capturing the output twice, once in OpenCV and once in Cocoa? If that works, that's good.
In some future sprint, if we want to show what OpenCV is extracting from the image, this will require a little refactoring, but that's ok.

File: Lighthouse Camera/UI/Main/PreviewView.swift [hunk 6]
```diff
@@ -0,0 +1,98 @@
...
+
+    self.session = captureSession
+
+    let sessionOutput = AVCapturePhotoOutput()
+    let backCamera = AVCaptureDevice.defaultDevice(withMediaType: AVMediaTypeVideo)
+
+    var error: NSError?
+    var input: AVCaptureDeviceInput!
+    do {
```
Comment (Yoric) [inline Lighthouse Camera/UI/Main/PreviewView.swift]: That looks weird. I guess you do that just to avoid capturing too many errors, right? Could you add a comment?

File: Lighthouse Camera/UI/Main/PreviewView.swift [hunk 7]
```diff
@@ -0,0 +1,98 @@
...
+      input = nil
+    }
+
+    if error == nil && captureSession.canAddInput(input) {
+      captureSession.addInput(input)
+
+      if captureSession.canAddOutput(sessionOutput) {
+        captureSession.addOutput(sessionOutput)
+
+        self.videoPreviewLayer.videoGravity = AVLayerVideoGravityResize;
```
Comment (Yoric) [inline Lighthouse Camera/UI/Main/PreviewView.swift]: Out of curiosity, what's `videoGravity`?

File: Lighthouse Camera/UI/Main/TTSHelper.swift [hunk 8]
```diff
@@ -0,0 +1,30 @@
+//
+//  TTSHelper.swift
```
Comment (Yoric) [inline Lighthouse Camera/UI/Main/TTSHelper.swift]: Wait, a TTSHelper and FTE in the same patch? Did you merge two patches accidentally?

File: Lighthouse Camera/UI/Main/ViewController.swift [hunk 9]
```diff
@@ -82,6 +80,7 @@ class ViewController: UIViewController {
   public dynamic func showFrame(frame: UIImage) {
     NSLog("ViewController.showFrame %f x %f\n", frame.size.width * frame.scale, frame.size.height * frame.scale);
     self.imageView.image = frame;
+    //AFter 3 seconds display previewView
```
Comment (Yoric) [inline Lighthouse Camera/UI/Main/ViewController.swift]: I don't understand this comment.

File: Lighthouse Camera/UI/Main/ViewController.swift [hunk 10]
```diff
@@ -111,6 +110,9 @@ class ViewController: UIViewController {
     self.isBusy = false;
     sViewController = self;
 
+    // Initialize the delegate to self indicating viewcontroller is going to implemenent PreviewViewDelegate
+    self.previewView.delegate = self
```
Comment (Yoric) [inline Lighthouse Camera/UI/Main/ViewController.swift]: Mmmh... Creating cycles is not a very good idea in Swift, as there is no garbage-collection. This is probably not going to hurt us at runtime, since the view is never going to disappear, but I have the intuition that it might be problematic in some tests.

If you do this, can you break the cycle somewhere (and document that you have broken it)?

File: Lighthouse Camera/Storyboards/LaunchScreen.storyboard [hunk 11]
```diff
@@ -16,12 +20,21 @@
                     <view key="view" contentMode="scaleToFill" id="Ze5-6b-2t3">
                         <rect key="frame" x="0.0" y="0.0" width="375" height="667"/>
                         <autoresizingMask key="autoresizingMask" widthSizable="YES" heightSizable="YES"/>
+                        <subviews>
```
Comment (sfoster) [inline Lighthouse Camera/Storyboards/LaunchScreen.storyboard]: The UX we are working towards has the preview and camera capture as the single point of entry for the app. So preview *is* the landing/home screen for the app (we might revise that based on user feedback)

File: Lighthouse Camera/UI/Main/AlertHelper.swift (hunk 1 above)
Comment (sfoster) [inline Lighthouse Camera/UI/Main/AlertHelper.swift]: I don't know if there's consensus on this stuff, but AFAICT this is the recommended formatting for switch statements in Swift. E.g. https://github.com/ochococo/Code-Formatting-In-Swift#control-operator-switch

File: Lighthouse Camera/UI/Main/ViewController.swift (hunk 9 above)
Comment (punamdahiya) [inline Lighthouse Camera/UI/Main/ViewController.swift]: yes, redundant will remove

File: Lighthouse Camera/UI/Main/TTSHelper.swift (hunk 8 above)
Comment (sfoster) [inline Lighthouse Camera/UI/Main/TTSHelper.swift]: I'm not convinced we should even speak of FTE as a thing - we don't have a demonstrated need for it yet. We talked about possibly having additional info - like verbose labels/descriptions which become more terse once the user has the hang of it.

Comment (punamdahiya) [inline Lighthouse Camera/UI/Main/TTSHelper.swift]: This was part of refactor to bring out code from view controller in separate files, but you are right it can be a separate patch

File: Lighthouse Camera/UI/Main/PreviewView.swift [hunk 12]
```diff
@@ -0,0 +1,98 @@
...
+      return true
+    }
+    set {
+      self.isAccessibilityElement = true
+    }
+  }
+    
+  override var accessibilityLabel: String? {
+    get {
+      return NSLocalizedString("Peview Screen", comment: "")
```
Comment (sfoster) [inline Lighthouse Camera/UI/Main/PreviewView.swift]: Typo s/Peview/Preview/. I'm not sure "Preview Screen" will be meaningful as a label, but we can adjust and iterate as we test it out

File: Lighthouse Camera/UI/Main/PreviewView.swift (hunk 6 above)
Comment (sfoster) [inline Lighthouse Camera/UI/Main/PreviewView.swift]: Yeah I've seen this as a couple nested blocks looping over devices, finding the back camera, trying to addInput, addOutput. This looks good to me.

File: Lighthouse Camera/UI/Main/AlertHelper.swift (hunk 1 above)
Comment (punamdahiya) [inline Lighthouse Camera/UI/Main/AlertHelper.swift]: That's what I saw too, swift has its own way of indenting switch case

File: Lighthouse Camera/UI/Main/PreviewView.swift (hunk 7 above)
Comment (sfoster) [inline Lighthouse Camera/UI/Main/PreviewView.swift]: https://developer.apple.com/reference/avfoundation/avcapturevideopreviewlayer/1386708-videogravity - its to do with how the video gets mapped/oriented into the preview layer rectangle. Weird name.

File: Lighthouse Camera/UI/Main/TTSHelper.swift [hunk 13]
```diff
@@ -0,0 +1,30 @@
...
+      case .AppReady: return NSLocalizedString("Lighthouse App is Ready", comment: "")
+      case .AppClosed: return NSLocalizedString("Lighthouse App is closing", comment: "")
+      }
+    }
+  }
+
+  @discardableResult convenience init(textID: TextStrings, rate: Float) {
+    self.init()
+    let utterance = AVSpeechUtterance(string: textID.description)
+    utterance.rate = rate
```
Comment (sfoster) [inline Lighthouse Camera/UI/Main/TTSHelper.swift]: This is a great start to build on, and we should land as-is. But FYI I did notice the TTS spoke slower than voiceover. I wonder if we can get the speaking rate from Voiceover and use the same when we call into here?

**Comment by punamdahiya** [comment]
Tried running app on simulator, there are no errors except camera preview doesn't work as simulator can't access camera

File: Lighthouse Camera/Storyboards/LaunchScreen.storyboard (hunk 11 above)
Comment (Yoric) [inline Lighthouse Camera/Storyboards/LaunchScreen.storyboard]: Correct me if I'm wrong, but I had the impression that `LaunchScreen` is displayed while the app is loading, right? If so, I don't quite understand this part of the patch.

File: Lighthouse Camera/UI/Main/AlertHelper.swift (hunk 1 above)
Comment (Yoric) [inline Lighthouse Camera/UI/Main/AlertHelper.swift]: Ok, fine by me.

File: Lighthouse Camera/UI/Main/PreviewView.swift (hunk 6 above)
Comment (Yoric) [inline Lighthouse Camera/UI/Main/PreviewView.swift]: I have no problem with the code, I just prefer it with more comments :)

**Comment by Yoric** [comment]
> Tried running app on simulator, there are no errors except camera preview doesn't work as simulator can't access camera

Oh, good. In this case, I have no problem. Just to be sure, can you run the `Lighthouse CameraUITests` on your machine? If not, we'll need to decide whether:

1. this patch needs to deactivate itself on the simulator; or
2. we sould only run the UITests with a device.

File: Lighthouse Camera/Storyboards/LaunchScreen.storyboard (hunk 11 above)
Comment (punamdahiya) [inline Lighthouse Camera/Storyboards/LaunchScreen.storyboard]: That's correct, launch screen is what first appears when the user taps the app icon before the app is finished launching, xcode now by default create a launchscreen.storyboard for every new project, my only change here is adding a UIImage and assign lighthouse image resource file to it. It displays lighhouse image instead of a blank screen while app is trying to load.
 We might need to build on our launch screen (wireframe 1 in UX spec) depending on more details from UX. 
Main storyboard is what is called when the app is running and handles all the logic.

File: Lighthouse Camera/Storyboards/LaunchScreen.storyboard [hunk 14]
```diff
@@ -1,7 +1,11 @@
...
-<document type="com.apple.InterfaceBuilder3.CocoaTouch.Storyboard.XIB" version="3.0" toolsVersion="11134" systemVersion="15F34" targetRuntime="iOS.CocoaTouch" propertyAccessControl="none" useAutolayout="YES" launchScreen="YES" useTraitCollections="YES" colorMatched="YES" initialViewController="01J-lp-oVM">
+<?xml version="1.0" encoding="UTF-8"?>
+<document type="com.apple.InterfaceBuilder3.CocoaTouch.Storyboard.XIB" version="3.0" toolsVersion="11762" systemVersion="16A323" targetRuntime="iOS.CocoaTouch" propertyAccessControl="none" useAutolayout="YES" launchScreen="YES" useTraitCollections="YES" colorMatched="YES" initialViewController="01J-lp-oVM">
+    <device id="retina4_7" orientation="portrait">
+        <adaptation id="fullscreen"/>
+    </device>
     <dependencies>
-        <plugIn identifier="com.apple.InterfaceBuilder.IBCocoaTouchPlugin" version="11106"/>
+        <deployment identifier="iOS"/>
+        <plugIn identifier="com.apple.InterfaceBuilder.IBCocoaTouchPlugin" version="11757"/>
```
Comment (punamdahiya) [inline Lighthouse Camera/Storyboards/LaunchScreen.storyboard]: Glad you asked , now I know what it does, found below answer

That plugin identifier tells Xcode that any Interface Builder Document (Storyboard or XIB) was built with Xcode, and the specific version of Xcode/Interface Builder that the interface was created with.

File: Lighthouse Camera/Storyboards/Main.storyboard [hunk 15]
```diff
@@ -18,54 +18,54 @@
...
                     </layoutGuides>
                     <view key="view" contentMode="scaleToFill" id="8bC-Xf-vdC">
-                        <rect key="frame" x="0.0" y="0.0" width="320" height="568"/>
+                        <rect key="frame" x="0.0" y="0.0" width="375" height="667"/>
                         <autoresizingMask key="autoresizingMask" widthSizable="YES" flexibleMinY="YES" heightSizable="YES" flexibleMaxY="YES"/>
                         <subviews>
-                            <imageView userInteractionEnabled="NO" contentMode="scaleAspectFill" horizontalHuggingPriority="251" verticalHuggingPriority="251" fixedFrame="YES" image="lighthouse" translatesAutoresizingMaskIntoConstraints="NO" id="BkF-XD-pqK">
-                                <rect key="frame" x="14" y="224" width="293" height="53"/>
-                                <autoresizingMask key="autoresizingMask" flexibleMaxX="YES" flexibleMaxY="YES"/>
+                            <imageView hidden="YES" userInteractionEnabled="NO" contentMode="scaleAspectFit" horizontalHuggingPriority="251" verticalHuggingPriority="251" fixedFrame="YES" translatesAutoresizingMaskIntoConstraints="NO" id="BkF-XD-pqK">
```
Comment (punamdahiya) [inline Lighthouse Camera/Storyboards/Main.storyboard]: We don't need to, will remove, we are handling now it in resetview inside main viewcontroller which switches between previewView and image captured view depending on where the app logic is

File: Lighthouse Camera/UI/Main/AlertHelper.swift (hunk 2 above)
Comment (punamdahiya) [inline Lighthouse Camera/UI/Main/AlertHelper.swift]: Done

File: Lighthouse Camera/UI/Main/PreviewView.swift (hunk 4 above)
Comment (punamdahiya) [inline Lighthouse Camera/UI/Main/PreviewView.swift]: Displays what the camera lens sees  in our UIView, previewView so that we should see a live "video" stream of phone camera's input within previewView

File: Lighthouse Camera/UI/Main/PreviewView.swift (hunk 5 above)
Comment (punamdahiya) [inline Lighthouse Camera/UI/Main/PreviewView.swift]: PreviewView class is handling primarily what camera is viewing live , may be I should call it LivePreviewView, yes we do need to figure OpenCV interactions in future sprints

Comment (punamdahiya) [inline Lighthouse Camera/UI/Main/PreviewView.swift]: Tried LivePreviewView too long, keeping PreviewView for now

File: Lighthouse Camera/UI/Main/PreviewView.swift (hunk 6 above)
Comment (punamdahiya) [inline Lighthouse Camera/UI/Main/PreviewView.swift]: Done

File: Lighthouse Camera/UI/Main/ViewController.swift (hunk 10 above)
Comment (punamdahiya) [inline Lighthouse Camera/UI/Main/ViewController.swift]: Explicitly declaring delegate variable inside PreviewView class to be weak so that it holds weak reference to view controller and will be deallocated  See https://www.natashatherobot.com/ios-weak-delegates-swift/

Comment (punamdahiya) [inline Lighthouse Camera/UI/Main/ViewController.swift]: will update this change

File: Lighthouse Camera/UI/Main/PreviewView.swift (hunk 12 above)
Comment (punamdahiya) [inline Lighthouse Camera/UI/Main/PreviewView.swift]: Done

File: Lighthouse Camera/UI/Main/TTSHelper.swift (hunk 13 above)
Comment (punamdahiya) [inline Lighthouse Camera/UI/Main/TTSHelper.swift]: Increased rate to 0.55 not sure if its the exact same as voice over but seems close.

File: Lighthouse Camera/Storyboards/LaunchScreen.storyboard (hunk 11 above)
Comment (Yoric) [inline Lighthouse Camera/Storyboards/LaunchScreen.storyboard]: Ah, my bad, got it.

File: Lighthouse Camera/UI/Main/PreviewView.swift [hunk 16]
```diff
@@ -0,0 +1,118 @@
...
+
+protocol PreviewViewDelegate: class {
+  func displayPreviewSession()
+}
+
+// Displays what the camera lens sees in UIView, previewView to
+// see a live video stream of phone camera's input within previewView
+
+class PreviewView: UIView {
+  // This delegate refers to a reference type and that class will be weakly retained
```
Comment (Yoric) [inline Lighthouse Camera/UI/Main/PreviewView.swift]: Nit: I don't think that this comment adds anything to the line below. Could you rather explain *why* it's a `weak`?

Comment (punamdahiya) [inline Lighthouse Camera/UI/Main/PreviewView.swift]: Updated by clarifying comment

**Comment by punamdahiya** [comment]
Thanks @Yoric , @sfoster for review, am able to successfully run Lighthouse CameraUITests on my machine , merging PR with master

**Comment by punamdahiya** [comment]
https://github.com/project-lighthouse/camera-app.ios/commit/fc18019cf9b69d102a27e1305abec6428fa4a435
